src/datatable/datatable.jl

```diff
@@ -74,32 +74,41 @@ type DataTable <: AbstractDataTable
     colindex::Index
 
     function DataTable(columns::Vector{Any}, colindex::Index)
-        ncols = length(columns)
-        if ncols > 1
-            nrows = length(columns[1])
-            equallengths = true
-            for i in 2:ncols
-                equallengths &= length(columns[i]) == nrows
-            end
-            if !equallengths
-                msg = "All columns in a DataTable must be the same length"
-                throw(ArgumentError(msg))
-            end
+        if length(columns) == length(colindex) == 0
+            return new(Vector{Any}(0), Index())
+        elseif length(columns) != length(colindex)
+            throw(DimensionMismatch("Number of columns and column names are different"))
         end
-        if length(colindex) != ncols
-            msg = "Columns and column index must be the same length"
-            throw(ArgumentError(msg))
+        lengths = length.(columns)
+        minlen, maxlen = extrema(lengths)
+        if minlen == 0 && maxlen == 0
+            return new(columns, colindex)
+        elseif (minlen == 0 && maxlen > 0) || any(x -> x != 0, mod(maxlen, lengths))
+            throw(DimensionMismatch("Incompatible lengths of arguments"))
+        else
+            for i in 1:length(columns)
+                if isa(columns[i], Range)
+                    columns[i] = collect(columns[i])
+                end
+                repeats = div(maxlen, length(columns[i]))
```

Comment:
The constructors seem like they are already pretty forgiving given that recycling is attempted at all. It would be an interesting duality of stances: forgiving and friendly during the construction of a DataTable but enforce strict behavior rules during use. We've taken a pretty clear stance on the latter by deciding no to allowing `dt[row, :col] = Nullabe()` to autopromote to a NullableArray and accept the assignment. I think there's precedent for the recycling behavior too. I'd like some additional perspectives here.